### PR TITLE
Implement manual state machine

### PR DESCRIPTION
**Pull request by rynowak**
This change is a rewrite of the action invoker into a manual state machine. This makes techempower plaintext go from 245krps -> 265krps.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 1]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
+                    {
+                        var filter = (IAsyncResourceFilter)state;
+                        if (_resourceExecutedContext == null)
+                        {
+                            // If we get here then the filter didn't call 'next' indicating a short circuit
+                            Debug.Assert(_resourceExecutingContext.Result != null);
+                            _resourceExecutedContext = new ResourceExecutedContext(_resourceExecutingContext, _filters)
+                            {
+                                Canceled = true,
+                                Result = _resourceExecutingContext.Result, // Can be null
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Comment incorrect. Have just asserted that `Result` is non-`null`.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 2]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
-                    // We don't execute Result Filters around the result.
-                    Debug.Assert(_exceptionContext != null);
-                    if (_exceptionContext.Result != null)
+                case State.ResultInside:
                     {
-                        // This means that exception filters returned a result to 'handle' an error.
-                        // We're not interested in seeing the exception details since it was handled.
-                        await InvokeResultAsync(_exceptionContext.Result);
+                        // The empty result is always flowed back as the 'executed' result
+                        var result = _resultExecutingContext?.Result ?? _result ?? new EmptyResult();
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: nit: The local can be inlined

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 3]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
+                        {
+                            if (_resourceExecutingContext == null)
+                            {
+                                _resourceExecutingContext = new ResourceExecutingContext(_controllerContext, _filters);
+                            }
+
+                            state = current.Filter;
+                            goto case State.ResourceSyncBegin;
+                        }
+                        else if (scope == Scope.Resource)
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: It would be nice to describe the transition here. For example, "All resource filters are on the stack, so we move on to put exception filters on the stack" but with better wording

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 4]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
+                        {
+                            Debug.Assert(_resourceExecutingContext != null);
+                            goto case State.ResourceInside;
+                        }
+                        else
+                        {
+                            Debug.Assert(scope == Scope.Invoker);
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Similar comment here, something along the lines of "there were no resource filters to invoke"

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 5]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
+
+                        _diagnosticSource.AfterOnResourceExecuted(resourceExecutedContext, filter);
+
+                        goto case State.ResourceEnd;
+                    }
+
+                case State.ResourceShortCircuit:
                     {
-                        await InvokeResourceFilterAsync();
                         Debug.Assert(_resourceExecutedContext != null);
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Debug.Assert(_resourceExecutedContext.Result != null)

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 6]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
+                case State.ResourceEnd:
+                    {
+                        if (scope == Scope.Resource)
+                        {
+                            isCompleted = true;
+                            return TaskCache.CompletedTask;
+                        }
+
+                        Debug.Assert(scope == Scope.Invoker);
+                        Rethrow(_resourceExecutedContext);
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: This makes it like it rethrows an exception unconditionally, it will be nice to be a bit more verbose and change the name to reflect that it will only throw if the previous scope ended in an unhandled exception. Something like RethrowIfUnhandledException

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 7]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
+                        {
+                            next = State.ExceptionAsyncResume;
+                            return task;
+                        }
+
+                        goto case State.ExceptionAsyncResume;
+                    }
+
+                case State.ExceptionAsyncResume:
+                    {
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Debug.Assert(_exceptionContext != null);

Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: It would be nice to describe the possible scenarios related to exceptionContext here in a comment. exceptionContext == null -> no exception happened. exceptionContext != null && exceptionContext.Exception -> an exception happened but was handled. exceptionContext != null && exceptionContext.Exception == null -> An exception happened but was handled by an inner exception handler.

I'm making all of these kind of comments because with the current implementation of the invoker it's really hard to grasp these details.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 8]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
+                    }
+
+                case State.ExceptionAsyncEnd:
+                    {
+                        var filter = (IAsyncExceptionFilter)state;
+                        var exceptionContext = _exceptionContext;
+
+                        _diagnosticSource.AfterOnExceptionAsync(exceptionContext, filter);
+
+                        if (exceptionContext.Exception == null)
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: exceptionContext can be null here if no exceptions happen, can't it? If so, can you add a test to cover this case?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 9]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
+                            goto case State.ExceptionAsyncEnd;
+                        }
+
+                        goto case State.ExceptionEnd;
+                    }
+
+                case State.ExceptionAsyncEnd:
+                    {
+                        var filter = (IAsyncExceptionFilter)state;
+                        var exceptionContext = _exceptionContext;
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Debug.Assert(_exceptionContext != null)

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 10]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
+                        if (task.Status != TaskStatus.RanToCompletion)
+                        {
+                            next = State.InvokeEnd;
+                            return task;
+                        }
+
+                        goto case State.ResourceEnd;
+                    }
+
+                case State.ExceptionEnd:
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Need to think what the right behavior is when an exception was handled by short-circuiting (setting Exception = null) but no result was set. (Need to take into account old MVC behavior)

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 11]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
+                        {
+                            goto case State.ActionInside;
+                        }
+                    }
+
+                case State.ActionAsyncBegin:
+                    {
+                        Debug.Assert(_actionExecutingContext != null);
+
+                        var filter = (IAsyncActionFilter)state;
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Do we want to also do Debug.Assert(state != null or state = _cursor.CurrentFilter.AsyncFilter) (applicable to other filter stages). The goal is to ensure that we always set the state to the filter in State.ActionNext

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 12]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
+                            };
+                        }
+
+                        _diagnosticSource.AfterOnActionExecution(_actionExecutedContext, filter);
+
+                        goto case State.ActionEnd;
+                    }
+
+                case State.ActionSyncBegin:
+                    {
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Debug.Assert(_actionExecutingContext != null)

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 13]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
+                            goto case State.ResultSyncBegin;
+                        }
+                        else
+                        {
+                            goto case State.ResultInside;
+                        }
+                    }
+
+                case State.ResultAsyncBegin:
+                    {
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Debug.Assert(_resultExecutingContext != null)

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 14]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
+                            {
+                                Canceled = true,
+                            };
+                        }
+
+                        _diagnosticSource.AfterOnResultExecution(_resultExecutedContext, filter);
+                        goto case State.ResultEnd;
+                    }
+
+                case State.ResultSyncBegin:
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Debug.Assert(_resultExecutingContext != null)

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 15]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
-                        };
+                            next = State.ResultEnd;
+                            return task;
+                        }
+
+                        goto case State.ResultEnd;
                     }
-                    else if (_exceptionContext.Exception != null)
+
+                case State.ResultEnd:
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: It would be good to add Debug.Assert(scope != <<scope>>) for some of these states, like in this case maybe Debug.Assert(scope != Scope.Exception). Also on the parts where the stack gets "Winded" and "Unwinded" its worth pointing out what happens with the state transitions, basically what the scenario is that we are covering on each of the branches. i.e, for scope == Scope.Resource indicate that the action being invoked contained resource filters and that we are unwinding those now.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 16]
```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
...
-                    _exceptionContext.ExceptionDispatchInfo = ExceptionDispatchInfo.Capture(exception);
-                }
+                    ExceptionDispatchInfo = ExceptionDispatchInfo.Capture(exception),
+                };
             }
         }
 
-        private async Task InvokeAllActionFiltersAsync()
+        private async Task InvokeNextActionFilterAsync()
         {
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Debug.Assert(_actionExecutingContext != null);

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 17]
```diff
@@ -483,219 +1112,132 @@ private Task InvokeAllExceptionFiltersAsync()
...
+        private async Task InvokeNextResultFilterAsync()
         {
-            _cursor.Reset();
-
-            _resultExecutingContext = new ResultExecutingContext(_controllerContext, _filters, result, _controller);
-            await InvokeResultFilterAsync();
-
-            Debug.Assert(_resultExecutingContext != null);
-            if (_resultExecutedContext.Exception != null && !_resultExecutedContext.ExceptionHandled)
+            try
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Debug.Assert(_resultExecutingContext != null);

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 18]
```diff
@@ -709,125 +1251,177 @@ private Task InvokeAllExceptionFiltersAsync()
...
+            if (context.ExceptionHandled)
+            {
+                return;
+            }
 
-                    if (_resultExecutingContext.Cancel == true)
-                    {
-                        // Short-circuited by setting Cancel == true
-                        _logger.ResourceFilterShortCircuited(item.Filter);
+            if (context.ExceptionDispatchInfo != null)
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: We only check for context.Exception in many of the states, is there a condition that when Exception != null ExceptionDispatchInfo != null too? If not, should we do a different check on the places where we check for Exception? My understanding is that ExceptionDispatchInfo is for async cases where Exception is for Sync cases.

**Comment by javiercn** [comment]
:watch:

**Comment by javiercn** [comment]
Overall looks good. Functionally everything looks alright compared to the old code (functionally equivalent). There are some small things we can improve (more Debug.Assert) and some comments at certain points of the state transitions to clarify what the algorithm does, as its hard to follow from the state transitions

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Fixed

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: I decided to fix this and added some comments to describe the possibilities.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: sure

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: sure

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs (hunk 6 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: won't fix, I think it's clear enough.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs (hunk 7 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: sure

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs (hunk 10 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: When you do this, we treat it as a short-circuit, and flow out the `EmptyResult` as the result.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs (hunk 11 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Added a bunch of these asserts.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs (hunk 15 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Please take a look at what's there and add specific comments. I've added comments in a few places already so if you think something specific is missing point it out.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs (hunk 18 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: `context.Exception` cannot be null when `context.ExceptionDispatchInfo` is non-null, so it covers all possibilities.

**Comment by rynowak** [comment]
@javiercn updated

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 19]
```diff
@@ -510,219 +1197,133 @@ private Task InvokeAllExceptionFiltersAsync()
...
-            await InvokeActionFilterAsync();
+            await InvokeNextActionFilterAsync();
 
             Debug.Assert(_actionExecutedContext != null);
             return _actionExecutedContext;
         }
 
-        private async Task InvokeActionFilterAsync()
+        private async Task InvokeActionMethodAsync()
         {
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Does this include the fix for ```public object Action(){ return Content("Hello world")!}``` ?

**Comment by javiercn** [comment]
:shipit:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs (hunk 19 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: This is up to date as of when I updated the PR. If this is missing then I'll address it when I rebase.
